In [21]:
import pandas as pd
import numpy as np
import os, glob

filename = os.path.join(os.getcwd(), "WebPurchase", "purchase_records.txt")
df = pd.read_table(filename, header=None)
df.columns = ['ac_id','purchase_date']

In [22]:
df.head(10)

ac_id purchase_date
0  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       6/17/17
1  AC3c8d6ef5-098e-421f-aad4-a563817aa655       6/17/17
2  AC916c16ac-344c-4654-b461-0994faf15c93       6/17/17
3  ACfd2bc854-d75c-400f-b96e-c4930e876c0f       6/17/17
4  ACe9aa7f4e-0e40-454e-883f-b0adb2ec0055       6/17/17
5  ACf00c42d0-b71f-4ec5-bdeb-697e36eece24       6/17/17
6  AC08b4867c-6097-4527-afa6-2113dceb1fd4       6/17/17
7  AC43c46d79-e344-4498-ba2b-bd175f586128       6/17/17
8  ACde9adb6e-3030-4899-9247-0d9d9a95f2b5       6/17/17
9  ACbf8bdcac-10cf-4905-9748-38485c7da428       6/17/17

In [23]:
filename = os.path.join(os.getcwd(), "WebPurchase", "webvisit.txt")
df1 = pd.read_table(filename, header=None)
df1.columns = ['ac_id','webvisited_date','visited_page']
df1.head(10)

ac_id webvisited_date   visited_page
0  AC6ebf5f05-f229-43d4-a810-2a8492180127         6/17/17               
1  ACce4cd8ea-8b3d-4767-9609-b353d1f633c0         6/17/17               
2  AC7d351059-47ef-4011-91d6-1ad803204854         6/17/17      MYACCOUNT
3  AC763265f0-40ac-400f-ad11-897ccdeb7d23         6/17/17        REWARDS
4  ACd89983be-b4e9-4e84-8936-f055f4ed3e7d         6/17/17  PRICINGENGINE
5  AC8ee59710-eb2f-4b48-a5c5-ae1627aba281         6/17/17  PRICINGENGINE
6  AC61169457-6279-428a-8414-ae968d891aa8         6/17/17      MYACCOUNT
7  AC2182f0dd-61f5-4f07-8144-44330b16a940         6/17/17  PRICINGENGINE
8  AC78b5ae75-a46c-4d84-a781-e29ba248ac52         6/17/17      MYACCOUNT
9  ACf5de13a2-16fa-4afe-be49-1d49c6ef6640         6/17/17  PRICINGENGINE

In [24]:
df.shape

(7229674, 2)

In [25]:
df1.shape

(1000000, 3)

Purchase records (df) much larger, contains whether customer made any purchase during the year. 

Web visit logs (df1) the history of customers at pages, not purchases.

In [27]:
df1_grouped_id = df1.groupby(['ac_id']).size().reset_index(name = 'visits')
df1_grouped_id

ac_id  visits
0       AC00001abb-e40b-4a19-978d-cc6becf33489       2
1       AC000061e1-9f80-4765-8ad4-993d5ab83e39       2
2       AC000064d8-cb24-44c6-b4ed-0cafa86b8d3e       3
3       AC0000754d-ba7a-45ed-b9a1-629d0bb0060d       2
4       AC0000b6cd-f27e-4c73-8d97-3510b8192c84       1
...                                        ...     ...
603625  ACffff7058-c1a4-45ea-9e0e-ab2c55a1b1d0       2
603626  ACffff8281-2731-4676-8856-acf3a588cf73       1
603627  ACffff8d3a-da1a-40c9-b29e-d5ed32713d5d       1
603628  ACffffd50a-291a-4067-92ae-9f8e8c629a9e       2
603629                                   ac_id       1

[603630 rows x 2 columns]

In [28]:
df1_grouped_id = df1_grouped.drop(603629,axis=0)
df1_grouped_id

ac_id  visits
0       AC00001abb-e40b-4a19-978d-cc6becf33489       2
1       AC000061e1-9f80-4765-8ad4-993d5ab83e39       2
2       AC000064d8-cb24-44c6-b4ed-0cafa86b8d3e       3
3       AC0000754d-ba7a-45ed-b9a1-629d0bb0060d       2
4       AC0000b6cd-f27e-4c73-8d97-3510b8192c84       1
...                                        ...     ...
603624  ACffff64ed-5e67-40ec-817d-4a05c62413e9       3
603625  ACffff7058-c1a4-45ea-9e0e-ab2c55a1b1d0       2
603626  ACffff8281-2731-4676-8856-acf3a588cf73       1
603627  ACffff8d3a-da1a-40c9-b29e-d5ed32713d5d       1
603628  ACffffd50a-291a-4067-92ae-9f8e8c629a9e       2

[603629 rows x 2 columns]

In [29]:
from functools import reduce
dfs = [df,df1]
df2 = reduce(lambda  left,right: pd.merge(left,right,on=['ac_id'],
                                               how='outer'), dfs).fillna('n/a')

In [30]:
df2.head(10)

ac_id purchase_date webvisited_date  \
0  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       6/17/17        11/16/17   
1  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       6/17/17         6/22/18   
2  AC6fe3b89d-437f-49ca-825a-b9d65f117f94        6/1/18        11/16/17   
3  AC6fe3b89d-437f-49ca-825a-b9d65f117f94        6/1/18         6/22/18   
4  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       2/19/18        11/16/17   
5  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       2/19/18         6/22/18   
6  AC6fe3b89d-437f-49ca-825a-b9d65f117f94        4/4/18        11/16/17   
7  AC6fe3b89d-437f-49ca-825a-b9d65f117f94        4/4/18         6/22/18   
8  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       8/22/17        11/16/17   
9  AC6fe3b89d-437f-49ca-825a-b9d65f117f94       8/22/17         6/22/18   

  visited_page  
0      REWARDS  
1      REWARDS  
2      REWARDS  
3      REWARDS  
4      REWARDS  
5      REWARDS  
6      REWARDS  
7      REWARDS  
8      REWARDS  
9      REWARDS

In [31]:
df2.shape

(10414416, 4)

10 million entries, multiple visits by customers, different purchase and web visit days. 

Feature to explore: possible correlation between the number of purchases and the number of website visits?

In [32]:
df2_grouped_id = df2.groupby(['ac_id']).size().reset_index(name = 'visit_count')
df2_grouped_id.head(10)

ac_id  visit_count
0  AC000017b9-3e7d-4133-9e59-ddddcb4a911d           13
1  AC00001abb-e40b-4a19-978d-cc6becf33489           16
2  AC000061e1-9f80-4765-8ad4-993d5ab83e39           10
3  AC000064d8-cb24-44c6-b4ed-0cafa86b8d3e           24
4  AC0000754d-ba7a-45ed-b9a1-629d0bb0060d           12
5  AC0000a060-4516-40d8-8e0a-066fd231cad7            3
6  AC0000a2fa-f6bf-4fc9-afb6-448f14bdd6b1            8
7  AC0000b6cd-f27e-4c73-8d97-3510b8192c84            7
8  AC0000bef9-0845-4c93-bec5-6c1e6d7fbe19            5
9  AC00011dd2-4a5e-4539-96ee-ac0a958245a4            7

In [33]:
df2_grouped_id.shape

(899927, 2)

Missing (900k-899927 = 73) entries being visited website but any specific pages, possibly stayed on home page

In [34]:
df2_grouped_page = df2.groupby(['visited_page']).size().reset_index(name = 'visit_count')
df2_grouped_page.head(10)

visited_page  visit_count
0                     2010240
1      MYACCOUNT      2002676
2  PRICINGENGINE      2015997
3        REWARDS      2005252
4            n/a      2380251

In [36]:
missing_sum = 2010240+2380251
missing_sum

4390491

Why are 4390491 occurances of visiting the site missing? Where are the customers visiting then? (Home page, bookmarked login, etc) Initial hypothesis that 73 entries were missing due to blank pages incorrect.

In [37]:
df2_grouped_id.tail(10)

ac_id  visit_count
899917  ACffff5419-8e25-44a4-baea-0df7d062912b            4
899918  ACffff59ab-65c9-420a-9280-a4e33d676004           11
899919  ACffff64ed-5e67-40ec-817d-4a05c62413e9           18
899920  ACffff7058-c1a4-45ea-9e0e-ab2c55a1b1d0           22
899921  ACffff8281-2731-4676-8856-acf3a588cf73           10
899922  ACffff8d20-b401-4bff-8ec0-e308c20f9822           10
899923  ACffff8d3a-da1a-40c9-b29e-d5ed32713d5d           10
899924  ACffffd50a-291a-4067-92ae-9f8e8c629a9e           12
899925  ACffffef1f-cfbc-4973-8926-35ef7a1491ab            5
899926                                   ac_id            8

In [39]:
df2_grouped_id = df2_grouped.drop(899926,axis=0)
df2_grouped_id

ac_id  visit_count
0       AC000017b9-3e7d-4133-9e59-ddddcb4a911d            1
1       AC00001abb-e40b-4a19-978d-cc6becf33489            1
2       AC000061e1-9f80-4765-8ad4-993d5ab83e39            1
3       AC000064d8-cb24-44c6-b4ed-0cafa86b8d3e            1
4       AC0000754d-ba7a-45ed-b9a1-629d0bb0060d            1
...                                        ...          ...
899921  ACffff8281-2731-4676-8856-acf3a588cf73            1
899922  ACffff8d20-b401-4bff-8ec0-e308c20f9822            1
899923  ACffff8d3a-da1a-40c9-b29e-d5ed32713d5d            1
899924  ACffffd50a-291a-4067-92ae-9f8e8c629a9e            1
899925  ACffffef1f-cfbc-4973-8926-35ef7a1491ab            1

[899926 rows x 2 columns]

Dropped last row, ac_id being just ac_id, visit count of 8. Remaining 74 entries of ac_id where?

In [41]:
df1.tail(10)

ac_id webvisited_date   visited_page
999990  AC616c65ce-3e64-4363-a330-4c474b27fe03         9/18/17      MYACCOUNT
999991  AC94814616-54bc-42ed-80dd-9c2b34ffc93f         9/18/17               
999992  AC2ccd897c-c800-42f4-b0a7-a1bb266ea6b9         9/18/17               
999993  AC386c9259-a3bb-450e-ab79-bfcccdff49dc         9/18/17               
999994  AC48ae0cb6-6852-4abe-8255-59f65d049ab4         9/18/17        REWARDS
999995  AC7b85dd54-70e2-4c5a-a1cd-66dd15280f81         9/18/17        REWARDS
999996  ACa9452768-3319-4ffc-9506-1ac344e0017c         9/18/17               
999997  ACc2db57f2-16cc-4593-8529-4890c37678ad         9/18/17        REWARDS
999998  AC93960bf7-a77a-4ff9-9c3a-9dc8cdba9d5a         9/18/17      MYACCOUNT
999999  ACc1913044-4541-4195-9d7a-5e2b138799e6         9/18/17  PRICINGENGINE

In [51]:
(df1['visited_page'].values == '').sum()

0

In [53]:
df1.dtypes

ac_id              object
webvisited_date    object
visited_page       object
dtype: object